tests/scripts/thread-cert/otci/otci/command_handlers.py

```diff
@@ -0,0 +1,211 @@
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+import logging
+import queue
+import re
+import threading
+import time
+from abc import abstractmethod
+from typing import Union, List, Pattern
+
+from .connectors import OtCliHandler
+from .errors import ExpectLineTimeoutError, CommandError
+from .utils import match_line
+
+
+class OTCommandHandler:
+    """This abstract class defines interfaces of a OT Command Handler."""
+
+    @abstractmethod
+    def execute_command(self, cmd: str, timeout: float) -> List[str]:
+        """Method execute_command should execute the OT CLI command within a timeout (in seconds) and return the
+        command output as a list of lines.
+
+        Note: each line SHOULD NOT contain '\r\n' at the end. The last line of output should be 'Done' or
+        'Error <code>: <msg>' following OT CLI conventions.
+        """
+        pass
+
+    @abstractmethod
+    def close(self):
+        """Method close should close the OT Command Handler."""
+        pass
+
+    @abstractmethod
+    def wait(self, duration: float) -> List[str]:
+        """Method wait should wait for a given duration and return the OT CLI output during this period.
+
+        Normally, OT CLI does not output when it's not executing any command. But OT CLI can also output
+        asynchronously in some cases (e.g. `Join Success` when Joiner joins successfully).s
+        """
+        pass
+
+
+class OtCliCommandRunner(OTCommandHandler):
+    __PATTERN_COMMAND_DONE_OR_ERROR = re.compile(
+        r'(Done|Error|Error \d+:.*|.*: command not found)$')  # "Error" for spinel-cli.py
+
+    __PATTERN_LOG_LINE = re.compile(r'((\[(NONE|CRIT|WARN|NOTE|INFO|DEBG)\])'
+                                    r'|(-.*-+: )'  # e.g. -CLI-----: 
+                                    r')')
+    """regex used to filter logs"""
+
+    __ASYNC_COMMANDS = {
+        'scan',
```

Comment:
`discover` is a todo.

`joiner start` outputs `Done` immediately, so it can be treated as a synchronized command.
Currently, `OTCI.joiner_start` does not return if the joining is successful or not, but returns immediately when `Done` is read (same as node.py).
I would prefer to enhance it in a future PR to wait for Join success or timeout.